Added workload fulltext docs (#40611)
ydb-platform/ydb: ydb/docs/en/core/reference/ydb-cli/workload-fulltext.md

|---|---|---|
| `--path` or `-p` | Path in the database where workload tables will be created. | `fulltext_workload` |

## Initializing the workload {#init}

Create the tables for the workload:

```bash
{{ ydb-cli }} workload fulltext init
```

### Available options {#init-options}

| Name | Description | Default value |
|---|---|---|
| `--min-partitions <value>` | Minimum number of table partitions. | `40` |
| `--partition-size <value>` | Target partition size, in MB. | `2000` |
| `--auto-partition <value>` | Enable auto-partitioning by load (`1` — enabled, `0` — disabled). | `1` |
| `--clear` | Drop and recreate the table if it already exists. | |

## Loading data {#load}

After initialization, load data into the table and build the fulltext index. There are two subcommands: `files` to import from an existing dataset and `generator` to generate synthetic data.

After import is complete, a fulltext index named `index` is automatically built on the `text` column.

### Importing from files {#load-files}

Import documents from files (CSV, TSV, or Parquet, optionally gzip-compressed). The dataset must contain `id` and `text` columns.

Example:

```bash

> {{ ydb-cli }} workload fulltext import files

Author: Параметр input имеет дефолтное значение, подразумевается минималистичный запуск на датасете который сразу содержит все нужные файлы с правильными именами.

> # Fulltext workload

Reviewer: ```suggestion
# Fulltext load
```

## Command structure {#structure}

```bash
{{ ydb-cli }} [global options...] workload fulltext [options...] <subcommand>
```

Subcommands:

```
fulltext            YDB fulltext workload
├─ init               Initialize tables for the workload
├─ import             Load data and build a fulltext index
│   ├─ files            Import data from files
│   └─ generator        Generate random text using a Markov chain model
├─ run                Run the workload
│   ├─ select           Search documents using fulltext queries
│   └─ upsert           Insert or update documents in the table
├─ clean              Drop tables created during initialization
└─ model              Build a Markov chain model from a text dataset
```

## Common command options {#common-options}

All commands support the following option:

| Name | Description | Default value |
|---|---|---|
| `--path` or `-p` | Path in the database where workload tables will be created. | `fulltext_workload` |

## Initializing the workload {#init}

> Create the table for the workload:

Reviewer: ```suggestion
Create the tables for the workload:
```

(или, если на самом деле одна — поправить RU и `Cleaning up`).

```bash

> {{ ydb-cli }} workload fulltext --path fulltext init

Reviewer: --path fulltext - остаток копипасты? В ru-версия такого флага нет

```suggestion
{{ ydb-cli }} workload fulltext init
```

Author: Да, это остатки от нейросети. Флаг есть, но его не обязательно указывать, там дефолт есть.

```

### Available options {#init-options}

| Name | Description | Default value |
|---|---|---|
| `--min-partitions <value>` | Minimum number of table partitions. | `40` |
| `--partition-size <value>` | Target partition size, in MB. | `2000` |
| `--auto-partition <value>` | Enable auto-partitioning by load (`1` — enabled, `0` — disabled). | `1` |
| `--clear` | Drop and recreate the table if it already exists. | |

## Loading data {#load}

After initialization, load data into the table and build the fulltext index. There are two subcommands: `files` to import from an existing dataset and `generator` to generate synthetic data.

After import is complete, a fulltext index is automatically built on the `text` column.

### Importing from files {#load-files}

Import documents from files (CSV, TSV, or Parquet, optionally gzip-compressed). The dataset must contain `id` and `text` columns.

Example:

```bash

> {{ ydb-cli }} workload fulltext import files

Reviewer: ```suggestion
After import is complete, a fulltext index named `index` is automatically built on the `text` column.
```

(И аналогично в RU стр. 64.)

{% include [run_options](./_includes/workload/run_options.md) %}

### Upsert workload {#run-upsert}

Continuously inserts new documents into the table using a Markov chain model to generate text.

```bash
{{ ydb-cli }} workload fulltext run upsert --model markov_dict.tsv.gz
```

#### Available options {#run-upsert-options}

| Name | Description | Default value |
|---|---|---|
| `--model <path>` or `-m <path>` | Path to the Markov chain model file (`.tsv.gz`). | Required |
| `--index-name <name>` | Name of the fulltext index to use. | `index` |
| `--bulk-size <value>` | Number of rows per upsert batch. | `100` |
| `--min-sentence-len <value>` | Minimum number of words in a generated document. | `100` |
| `--max-sentence-len <value>` | Maximum number of words in a generated document. | `1000` |

{% include [run_options](./_includes/workload/run_options.md) %}

## Building a Markov chain model {#model}

Before using the generator or the `run upsert` / `run select` modes with generated queries, you need to build a Markov chain model from a text dataset. The model captures word transition probabilities and is used to generate realistic text.

> ```bash

Reviewer: --order дефолт 1, но и пример выше, и Usage example используют --order 3. Скорее всего, надо объяснить пользователю

Author: Да, распишу по подробнее про то что такое order.

{{ ydb-cli }} workload fulltext model --input wikipedia_sample.csv.gz --output markov_dict.tsv.gz --order 3
```

### Available options {#model-options}

| Name | Description | Default value |
|---|---|---|
| `--input <path>` or `-i <path>` | Path to the dataset file or directory. Supports `.csv[.gz]` and `.tsv[.gz]` formats. The file must have a `text` column. | Required |
| `--output <path>` or `-o <path>` | Output file path for the model dictionary. | `markov_dict.tsv.gz` |
| `--order <value>` or `-n <value>` | Order of the Markov chain (n-gram context size). Order 1 uses unigram context, order 2 uses bigram context, etc. Must be between 1 and 5. | `1` |

## Cleaning up {#cleanup}

Drop all tables created during initialization:

```bash
{{ ydb-cli }} workload fulltext clean
```

> The command has no additional parameters.

Reviewer: В RU-версии аналогичной строки нет

Author: Видимо удалю, строка не имеет смысла.

## Usage examples {#examples}

### Example with a generated dataset

> 1a. Download a Markov chain model from s3:

Reviewer: Нумерация `1a.` / `1b.` — нестандартная для markdown и YFM может отрендерить как обычный текст вместо списка.

> 1a. Download a Markov chain model from s3:

Reviewer: ```suggestion
1a. Download a Markov chain model from S3:
```

    ```bash
    wget https://storage.yandexcloud.net/ydb-public/markov_dict.tsv.gz
    ```

> 1b. Build a Markov chain model from a Wikipedia sample:

Reviewer: см выше коммент, касающийся 1a и 1b

    ```python
    from datasets import load_dataset

    ds = load_dataset(
       "rumbleFTW/wikipedia-20220301-en-raw",
       split="train[:1000000]",
       streaming=False,
       )
    ds.to_csv('wikipedia_sample.csv.gz', compression='gzip', index=False)
    ```

    ```bash
    {{ ydb-cli }} workload fulltext model --input wikipedia_sample.csv.gz --output markov_dict.tsv.gz --order 3
    ```

2. Initialize the workload table:

    ```bash
    {{ ydb-cli }} workload fulltext init
    ```

3. Generate and load synthetic documents:

    ```bash
    {{ ydb-cli }} workload fulltext import generator
    ```

4. Run the search workload:

    ```bash
    {{ ydb-cli }} workload fulltext run select
    ```

5. Run the upsert workload:

    ```bash
    {{ ydb-cli }} workload fulltext run upsert
    ```

6. Clean up:

    ```bash
    {{ ydb-cli }} workload fulltext clean
    ```

### Example with the MS MARCO dataset

> 1. Download the quality bundle (contains `documents.tsv.gz`, `queries.tsv.gz`, `markov_dict.tsv.gz`, and 'query_relevances.tsv.gz'):

Reviewer: ```suggestion
1. Download the quality bundle (contains `documents.tsv.gz`, `queries.tsv.gz`, `markov_dict.tsv.gz`, and `query_relevances.tsv.gz`):
```

## Initializing the workload {#init}

Create the tables for the workload:

```bash
{{ ydb-cli }} workload fulltext init
```

### Available options {#init-options}

| Name | Description | Default value |
|---|---|---|
| `--min-partitions <value>` | Minimum number of table partitions. | `40` |
| `--partition-size <value>` | Target partition size, in MB. | `2000` |
| `--auto-partition <value>` | Enable auto-partitioning by load (`1` — enabled, `0` — disabled). | `1` |
| `--clear` | Drop and recreate the table if it already exists. | |

## Loading data {#load}

After initialization, load data into the table and build the fulltext index. There are two subcommands: `files` to import from an existing dataset and `generator` to generate synthetic data.

After import is complete, a fulltext index named `index` is automatically built on the `text` column.

### Importing from files {#load-files}

Import documents from files (CSV, TSV, or Parquet, optionally gzip-compressed). The dataset must contain `id` and `text` columns. For column `id` used storage type Uint64 in internal table, for column `text` used type String.

Example:

```bash

> {{ ydb-cli }} workload fulltext import files

Author: Имена файлов совпадают с дефолтными значениями параметров и добавлять их явно не обязательно. Это часть дизайна.

```

#### Available options {#load-files-options}

| Name | Description | Default value |
|---|---|---|
| `--input <path>` or `-i <path>` | Path to the dataset file or directory. Supported formats: CSV/TSV (optionally gzip-compressed), Parquet. Only `id` and `text` columns are imported. | Required |

{% include [load_options](./_includes/workload/load_options.md) %}

> 

Author: Файлы написаны на том языке на котором нужно. Тут нет бага

    5           281 281     0       0       32      38      40      42
    6           278 278     0       0       32      39      41      43
    7           279 279     0       0       32      38      39      43
    8           275 275     0       0       33      39      42      46
    9           282 282     0       0       32      37      39      40
    10          293 293     0       0       31      37      39      41

    Total       Txs Txs/Sec Retries Errors  p50(ms) p95(ms) p99(ms) pMax(ms)
    10         2781 278.1   0       0       32      38      41      123
    ```
6. Clean up:

    ```bash
    {{ ydb-cli }} workload fulltext clean
    ```

### Example with the MS MARCO dataset

1. Download the quality bundle (contains `documents.tsv.gz`, `queries.tsv.gz`, `markov_dict.tsv.gz`, and `query_relevances.tsv.gz`):

    ```bash
    wget https://storage.yandexcloud.net/ydb-public/quality-bundle.tar
    tar -xf quality-bundle.tar
    ```

2. Initialize the workload table:

    ```bash
    {{ ydb-cli }} workload fulltext init
    ```

3. Import documents from the dataset:

    ```bash
    {{ ydb-cli }} workload fulltext import files
    ```

> 

Author: Не надо добавлять сюда флагов, все файлы подцепятся через дефолт параметров

4. Run the search workload using the pre-built queries:

    ```bash
    {{ ydb-cli }} workload fulltext run select --quality
    ```
    Output example:
    ```
    Search quality measurement...
    Search quality measurement completed for 100 queries in 1 seconds.
    nDCG@10:  0.270

>     Errors:   0

Author: Не надо добавлять дополнительных опций. Имена файлов совпадают с дефолтом всех параметров.
